core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
 class Transaction:
     def __init__(
         self,
-        inputs: list,
-        outputs: list,
-        keychain,
+        inputs: List[CardanoTxInputType],
+        outputs: List[CardanoTxOutputType],
+        keychain: seed.Keychain,
         protocol_magic: int,
-        input_coins_sum: int,
-    ):
+        fee: int,
+        ttl: int,
+    ) -> None:
         self.inputs = inputs
         self.outputs = outputs
         self.keychain = keychain
-        # attributes have to be always empty in current Cardano
-        self.attributes = {}
+        self.fee = fee
+        self.ttl = ttl
 
         self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic, "Unknown")
         self.protocol_magic = protocol_magic
-        self.input_coins_sum = input_coins_sum
 
-    def _process_outputs(self):
-        change_addresses = []
-        change_derivation_paths = []
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
+    def _build_inputs(self) -> list:
+        inputs_for_cbor = []
+        for input in self.inputs:
+            inputs_for_cbor.append([input.prev_hash, input.prev_index])
+
+        return inputs_for_cbor
+
+    def _build_outputs(self) -> list:
+        outputs_for_cbor = []
 
         for output in self.outputs:
+            amount = output.amount
             if output.address_n:
-                address, _ = derive_address_and_node(self.keychain, output.address_n)
-                change_addresses.append(address)
-                change_derivation_paths.append(output.address_n)
-                change_coins.append(output.amount)
+                address, _ = derive_address_and_node(
+                    self.keychain, output.address_n, self.protocol_magic
+                )
+                # set output address so it can later be shown to the user if needed
+                output.address = address
```

Comment:
this kind of smells - one would not expect a "build" function to mutate the class properties. Isn't there a way to achieve the same without this hack?

Reply:
This is needed just for the show_tx function. I wasn't able to come up with a better way. I can also add something like this to show_tx:
`        if not output.address:
            address, _ = derive_address_and_node(keychain, output.address_n)
        else:
            address = output.address

        await confirm_sending(ctx, output.amount, address)`
which would require sending in the keychain and deriving the address a second time. Or storing it in some property as it was before, but I wanted to simplify it.

Reply:
What about making `show_tx` part of the `Transaction` class? The class has all the means to compute the addresses in whatever format is necessary, and it also knows the inputs and outputs to show.

Reply:
To me it feels like showing the transaction is a different matter for which the Transaction class should not be responsible. Also, maybe we don't even the Transaction class anymore? (see comment below)